File: src/core/cooldowns.py
import asyncio
from time import time


class CooldownsManager:
    def __init__(self) -> None:
        self.cooldowns: dict[int, dict[str, float]] = {}

    def add_user(self, user_id: int, context):
        self.cooldowns[user_id] = {"last_usage": time()}
        asyncio.create_task(self.schedule_delete_user(user_id, context))

    def find_user(self, user_id):
        return self.cooldowns.get(user_id)

    async def delete_user(self, user_id, context):
        user = self.find_user(user_id=user_id)
        if user:
            del self.cooldowns[user_id]

            return await context.channel.send(
                content=f"{context.user.mention}, vous pouvez à nouveau jouer !",
            )
        raise Exception("error, no user deleted")

    async def schedule_delete_user(self, user_id, context):
        await asyncio.sleep(14400)
        await self.delete_user(user_id, context)

File: src/client/utils/cooldowns.ts
import { Collection } from "discord.js";
import { CustomChatInputCommandInteraction } from "../types/customInteraction";

type CooldownsConfig = {
  duration: number;
};

export const cooldownConfig: CooldownsConfig = {
  duration: 14_400_000,
};

export class Cooldowns {
  private declare cooldowns: Collection<string, number>;
  private declare config: CooldownsConfig;

  constructor(config: CooldownsConfig) {
    this.cooldowns = new Collection<string, number>();
    this.config = config;
  }

  addUser = (
    userId: string,
    interaction: CustomChatInputCommandInteraction
  ): void => {
    this.cooldowns.set(userId, Date.now());
    setTimeout(() => {
      this.cooldowns.delete(userId);
      if (!interaction.channel?.isSendable()) return;
      interaction.channel.send({
        content: `<@!${interaction.user.id}>, vous pouvez a nouveau jouer !`,
      });
    }, this.config.duration);
  };

  findUser = (userId: string): boolean => {
    return this.cooldowns.has(userId);
  };

  getRemainingTimeFor = (userId: string): number => {
    const userRemaningCooldown = this.cooldowns.get(userId);
    if (!userRemaningCooldown) {
      throw Error();
    }
    const now = Date.now();
    const timeSinceLastUsage = now - userRemaningCooldown;
    return this.config.duration - timeSinceLastUsage;
  };

  static formatCooldown = (cooldown: number): string => {
    return `${Math.floor(cooldown / 3_600_000)} heure(s) ${Math.floor(
      (cooldown % 3_600_000) / 60_000
    )} minute(s) ${Math.floor((cooldown % 60_000) / 1_000)} seconde(s)`;
  };
}
